### DISCUSSION TITLE
Apache Error 500 v23.01

### DISCUSSION THREAD

**Discussion by wade-lich** [opendcim/openDCIM#1365] [Q&A] [answered]
Hi all,

  While trying to install OpenDCIM 23.01 I encountered the following error. 

Fatal error: Uncaught Error: Undefined constant PDO::MYSQL_ATTR_INIT_COMMAND in /var/www/html/db.inc.php:34 Stack trace: #0 /var/www/html/index.php(11): require_once() #1 {main} thrown in /var/www/html/db.inc.php on line 34

The preflight had everything listed as GOOD.

rightscheck was correct.

I have verified the db user/password are valid from the command line.

PHP 8.0.20

Apache 2.4.37

Centos/RHEL 8.7

MariaDB 8.0.30

This is the code segment

 $pdo_options=array(
                PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8',
                PDO::ATTR_PERSISTENT => true
        );

        try{
                $pdoconnect="mysql:host=$dbhost;port=$dbport;dbname=$dbname";
                $dbh=@new PDO($pdoconnect,$dbuser,$dbpass,$pdo_options);
                $dbh->exec("SET @@SESSION.sql_mode = ''");
        }catch(PDOException $e){
                printf( "Error!  %s\n", $e->getMessage() );
                die();
        }

9 out 10 chance I missed a really simple step, but I'm not seeing the issue. 

Any guidance would be appreciated!

Regards,

  Wade

**Comment by wilpig** [comment]
It's most likely you didn't install the php-pdo module. See the rhel9 install instructions they should be similar enough to rocky/centos/alma/rhel 8.  If I need to make a specific set for 8 with the appstreams/module statements I can.

https://github.com/opendcim/openDCIM/wiki/RHEL


Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: wade-lich ***@***.***>
Sent: Thursday, June 1, 2023 5:32:11 AM
To: opendcim/openDCIM ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [opendcim/openDCIM] Apache Error 500 v23.01 (Discussion #1365)


Hi all,

While trying to install OpenDCIM 23.01 I encountered the following error.

Fatal error: Uncaught Error: Undefined constant PDO::MYSQL_ATTR_INIT_COMMAND in /var/www/html/db.inc.php:34 Stack trace: #0 /var/www/html/index.php(11): require_once() #1<https://github.com/opendcim/openDCIM/issues/1> {main} thrown in /var/www/html/db.inc.php on line 34

The preflight had everything listed as GOOD.

rightscheck was correct.

I have verified the db user/password are valid from the command line.

PHP 8.0.20

Apache 2.4.37

Centos/RHEL 8.7

MariaDB 8.0.30

This is the code segment

$pdo_options=array(
PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8',
PDO::ATTR_PERSISTENT => true
);

    try{
            $pdoconnect="mysql:host=$dbhost;port=$dbport;dbname=$dbname";
            ***@***.*** PDO($pdoconnect,$dbuser,$dbpass,$pdo_options);
            $dbh->exec("SET @@SESSION.sql_mode = ''");
    }catch(PDOException $e){
            printf( "Error!  %s\n", $e->getMessage() );
            die();
    }


9 out 10 chance I missed a really simple step, but I'm not seeing the issue.

Any guidance would be appreciated!

Regards,

Wade

—
Reply to this email directly, view it on GitHub<https://github.com/opendcim/openDCIM/discussions/1365>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAMTY6R2VT5FDKWE3TBKULDXJBVSXANCNFSM6AAAAAAYWWYSN4>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

[ { ***@***.***": "http://schema.org", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "https://github.com/opendcim/openDCIM/discussions/1365", "url": "https://github.com/opendcim/openDCIM/discussions/1365", "name": "View Discussion" }, "description": "View this Discussion on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

**Comment by wade-lich** [comment]
I verified it is installed.

Package httpd-2.4.37-51.0.1.module+el8.7.0+20778+02173b8e.x86_64 is already installed.
Package php-8.0.20-2.module+el8.7.0+20803+bb5ab920.x86_64 is already installed.
**Package php-pdo-8.0.20-2.module+el8.7.0+20803+bb5ab920.x86_64 is already installed.**
Package php-mbstring-8.0.20-2.module+el8.7.0+20803+bb5ab920.x86_64 is already installed.
Package php-snmp-8.0.20-2.module+el8.7.0+20803+bb5ab920.x86_64 is already installed.
Package php-xml-8.0.20-2.module+el8.7.0+20803+bb5ab920.x86_64 is already installed.
Package php-gd-8.0.20-2.module+el8.7.0+20803+bb5ab920.x86_64 is already installed.
Package graphviz-2.40.1-43.el8.x86_64 is already installed.
No match for argument: graphviz-gd - **I need to track this one down.**
Package wget-1.19.5-10.0.1.el8.x86_64 is already installed.
Package vim-enhanced-2:8.0.1763-19.0.1.el8_6.4.x86_64 is already installed.



-- Wade

**Comment by wilpig** [comment]
Did you restart apache and/or php-fpm after installing the php-pdo module?

**Comment by wade-lich** [comment]
many times and rebooted too.

**Comment by wilpig** [comment]
I will try to recreate.  I just installed a fresh RHEL8.7 system and will try to duplicate my dev environment there, but I have plans for the next week already that will keep me away from the keyboard so don't expect anything extremely quickly.

**Reply by wade-lich** [reply]
ACK

**Reply by wilpig** [reply, marked as answer]
https://github.com/opendcim/openDCIM/wiki/RHEL

RHEL8 required packages and fully tested instructions added.  This was done from a minimal install of RHEL8.7 same as I did for the RHEL9.   I didn't have the 8.8 ISO downloaded or I would have done that, it shouldn't have changed.  The differences with centos8 stream should be minimal.

**Comment by wade-lich** [comment]
php-mysqlnd was missing.  I had copied and pasted the install line from the link, but it didn't even attempt to install it.  So, I've moved past that part onto another problem... 
I'll try to save us both time by asking, can I do a direct upgrade from 20.01 -> 23.01?  Or,  do I need to hit the major release(s) in the middle?  If I can do a direct upgrade, then I have additional issues.  If I can't then I'll head down that rabbit hole.

**Reply by wilpig** [reply]
Yes a direct upgrade should work fine.

**Comment by wade-lich** [comment]
Ok then... Here's my newest error.
from: /rowview.php?row=20

Fatal error: Unparenthesized `a ? b : c ? d : e` is not supported. Use either `(a ? b : c) ? d : e` or `a ? b : (c ? d : e)` in /var/www/html/classes/Device.class.php on line 369

Additionally, when running asset reports, it loads the meter, says starting then does noting after that.

**Reply by samilliken** [reply]
Yeah, that one is going to have to be patched to take out the unary operation that wilpig is so fond of.   For now I'd just comment out that line with a //

As for the asset reports, nothing in the error logs?

**Comment by wade-lich** [comment]
Fatal error: Uncaught Error: Class "PhpOffice\PhpSpreadsheet\Worksheet\PageSetup" not found in /var/www/html/report_asset_Excel.php:171 Stack trace: #0 {main} thrown in /var/www/html/report_asset_Excel.php on line 171

**Reply by samilliken** [reply]
Ugh, where was all this testing when I had asked for folks to test?  LOL   Apparently that class/method doesn't even exist in the new PHPSpreadsheet library.   That report was contributed and abandoned.   I hate its very existence but I will likely dig into it and try to fix it or rewrite it.

**Reply by wade-lich** [reply]
I'll drill though all of this and compile a list.  Is there a way to enable error reporting globally?  I've been doing it on a file by file basis.  
putting it in the index.php and php.ini didn't do it.

ini_set('display_startup_errors', 1);
ini_set('display_errors', 1);
error_reporting(-1);

**Reply by samilliken** [reply]
Well, @wade-lich , I was able to find an error in report_asset_Excel, but not at the spot you reference.   There was a call to a deprecated function on line 1781, but removing lines 1781 and 1782 got rid of that.   I was able to produce a working report with that modification.   I can't reproduce the order you did list here about the PageSetup class not being found.

**Reply by wade-lich** [reply]
I forgot to allow for the insertion of the error reporting commands which added 3 lines. So it should be line 168.  It manifests itself in the lower left corner of the page.  See attached.
<img width="907" alt="Screen Shot 2023-06-02 at 8 00 51 AM" src="https://github.com/opendcim/openDCIM/assets/19672155/735ca18e-2e89-46d0-baf1-fc9f8d80b124">

also 

Fatal error: Uncaught Error: Class "PhpOffice\PhpSpreadsheet\Spreadsheet" not found in /var/www/html/report_projects.php:47 Stack trace: #0 {main} thrown in /var/www/html/report_projects.php on line 47

Fatal error: Unparenthesized `a ? b : c ? d : e` is not supported. Use either `(a ? b : c) ? d : e` or `a ? b : (c ? d : e)` in /var/www/html/classes/DevicePorts.class.php on line 185 -- comment it out?

**Reply by samilliken** [reply]
Your issue is looking more and more like something screwed up in your setup.   That class does exist and I just ran that report on the demo site.

**Reply by samilliken** [reply]
@wade-lich I don't see php-zip in the RHEL 9 wiki page that @wilpig made.   It has to be installed and enabled for PHPSpreadsheet.

**Reply by wade-lich** [reply]
The only php zip utility in  8's repo is php-pecl-zip.x86_64 : A ZIP archive management extension, which I do have installed.    php-zip does not show up in a repo search. Is this package acceptable?  

: [1347] root@dcopsvm-166:www # ; php --ini
Configuration File (php.ini) Path: /etc
Loaded Configuration File:         /etc/php.ini
Scan for additional .ini files in: /etc/php.d
Additional .ini files parsed:      /etc/php.d/10-opcache.ini,
/etc/php.d/20-bz2.ini,
/etc/php.d/20-calendar.ini,
/etc/php.d/20-ctype.ini,
/etc/php.d/20-curl.ini,
/etc/php.d/20-dom.ini,
/etc/php.d/20-exif.ini,
/etc/php.d/20-fileinfo.ini,
/etc/php.d/20-ftp.ini,
/etc/php.d/20-gd.ini,
/etc/php.d/20-gettext.ini,
/etc/php.d/20-iconv.ini,
/etc/php.d/20-intl.ini,
/etc/php.d/20-ldap.ini,
/etc/php.d/20-mbstring.ini,
/etc/php.d/20-mysqlnd.ini,
/etc/php.d/20-pdo.ini,
/etc/php.d/20-phar.ini,
/etc/php.d/20-simplexml.ini,
/etc/php.d/20-snmp.ini,
/etc/php.d/20-sockets.ini,
/etc/php.d/20-sqlite3.ini,
/etc/php.d/20-tokenizer.ini,
/etc/php.d/20-xml.ini,
/etc/php.d/20-xmlwriter.ini,
/etc/php.d/20-xsl.ini,
/etc/php.d/30-mysqli.ini,
/etc/php.d/30-pdo_mysql.ini,
/etc/php.d/30-pdo_sqlite.ini,
/etc/php.d/30-xmlreader.ini,
/etc/php.d/40-zip.ini


FWIW - I dumped the entire directory and started from scratch now that the DB has been upgraded from 20.01 -> 23.01.  That part took four refreshes before it completed.  I configured httpd for SSL, edited the db.inc.php, set my db/dbuser/dbpass, disable apache auth, enabled ldap, updated the install base URL, renamed install.php and migrated my images and drawings.  So far everything, except for some of the reports, is working just as before/expected.  Also had to rebuild the image cache.

I appreciate the help you're giving!!

**Comment by wilpig** [comment]
We discussed that module on IRC this morning and it appears the rhel build of PHP8 contains the zip module as part of the base

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by wade-lich** [comment]
Here's a complete list of the Reports that are causing 500 errors:

Fatal error: Uncaught Error: Undefined constant PhpOffice\PhpSpreadsheet\Collection\CellsFactory::cache_in_memory_serialized in /var/www/ip_default/report_asset_Excel.php:1781 Stack trace: #0 {main} thrown in /var/www/ip_default/report_asset_Excel.php on line 1781
-----------------------------------
Fatal error: Uncaught PhpOffice\PhpSpreadsheet\Exception: File /var/www/ip_default/logo.png not found! in /var/www/ip_default/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Worksheet/Drawing.php:134 Stack trace: #0 /var/www/ip_default/report_projects.php(67): PhpOffice\PhpSpreadsheet\Worksheet\Drawing->setPath() #1 {main} thrown in /var/www/ip_default/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Worksheet/Drawing.php on line 134
---------------------------------
report_network_map.php with "Cabinet" option:
This page contains the following errors:
error on line 2 at column 1: Extra content at the end of the document
Below is a rendering of the page up to the first error.
---------------------------
report_network_map.php with "Zone" option:

This page contains the following errors:
error on line 2 at column 1: Extra content at the end of the document
Below is a rendering of the page up to the first error.
-----------------------------
Fatal error: Uncaught Error: Undefined constant PhpOffice\PhpSpreadsheet\Collection\CellsFactory::cache_in_memory_serialized in /var/www/ip_default/report_vendor_model.php:96 Stack trace: #0 {main} thrown in /var/www/ip_default/report_vendor_model.php on line 96
--------------------------
report_outage_simulator.php with "Only show down/unknown devices" option:
(I think this one is more a validation issue.  Only happens when this is the only option chosen)

Fatal error: Uncaught TypeError: count(): Argument #1 ($value) must be of type Countable|array, null given in /var/www/ip_default/report_outage_simulator.php:166 Stack trace: #0 {main} thrown in /var/www/ip_default/report_outage_simulator.php on line 166

------------------------

Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Warning: Undefined property: PowerDistribution::$DeviceID in /var/www/ip_default/misc.inc.php on line 190
Fatal error: Uncaught PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Reservation' in 'where clause' in /var/www/ip_default/report_project_outage_simulator.php:411 Stack trace: #0 /var/www/ip_default/report_project_outage_simulator.php(411): PDOStatement->execute() #1 {main} thrown in /var/www/ip_default/report_project_outage_simulator.php on line 411

------------------
/report_panel_schedule.php - produces no output at all and no errors.

--------------------------------


Fatal error: Uncaught TypeError: Unsupported operand types: string * int in /var/www/ip_default/report_panel_schedule.php:197 Stack trace: #0 {main} thrown in /var/www/ip_default/report_panel_schedule.php on line 197

-------------------------------


Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 214
Warning: Undefined property: Device::$Reservation in /var/www/ip_default/report_cabinets.php on line 220

Fatal error: Uncaught Mpdf\MpdfException: Data has already been sent to output (/var/www/ip_default/report_cabinets.php at line 220), unable to output PDF file in /var/www/ip_default/vendor/mpdf/mpdf/src/Mpdf.php:9584 Stack trace: #0 /var/www/ip_default/template_mpdf_reports.inc.php(101): Mpdf\Mpdf->Output() #1 /var/www/ip_default/report_cabinets.php(424): include('...') #2 {main} thrown in /var/www/ip_default/vendor/mpdf/mpdf/src/Mpdf.php on line 9584

This happens with any number of cabinets selected
-----------------------------
Fatal error: Uncaught DivisionByZeroError: Division by zero in /var/www/ip_default/report_audit_frequency.php:357 Stack trace: #0 {main} thrown in /var/www/ip_default/report_audit_frequency.php on line 357
-----------------------------
Warning: Undefined array key 0 in /var/www/ip_default/report_surplus.php on line 21

Warning: Attempt to read property "Name" on null in /var/www/ip_default/report_surplus.php on line 21

Fatal error: Uncaught PhpOffice\PhpSpreadsheet\Exception: File /var/www/ip_default/logo.png not found! in /var/www/ip_default/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Worksheet/Drawing.php:134 Stack trace: #0 /var/www/ip_default/report_surplus.php(69): PhpOffice\PhpSpreadsheet\Worksheet\Drawing->setPath() #1 {main} thrown in /var/www/ip_default/vendor/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Worksheet/Drawing.php on line 134
------------------------------
Fatal error: Uncaught Error: Class "mPDF" not found in /var/www/ip_default/report_supply_status.php:17 Stack trace: #0 {main} thrown in /var/www/ip_default/report_supply_status.php on line 17
--------------------------------
This is a validation error I belive:

Fatal error: Uncaught PHPMailer\PHPMailer\Exception: Invalid address: (to): in /var/www/ip_default/vendor/phpmailer/phpmailer/src/PHPMailer.php:1094 Stack trace: #0 /var/www/ip_default/vendor/phpmailer/phpmailer/src/PHPMailer.php(1014): PHPMailer\PHPMailer\PHPMailer->addOrEnqueueAnAddress() #1 /var/www/ip_default/rackrequest.php(59): PHPMailer\PHPMailer\PHPMailer->addAddress() #2 {main} thrown in /var/www/ip_default/vendor/phpmailer/phpmailer/src/PHPMailer.php on line 1094





I'm at a loss where to go from here.

**Comment by wade-lich** [comment]
As it turns out, 80% of the problems went away when error reporting was removed.  
	//ini_set('display_errors', 1);
	//ini_set('display_startup_errors', 1);
	//error_reporting(E_ALL);
	
What it tells me is there's a lot of warning/errors/deprecation that the code just otherwise skips over unless it's asked to share.  At this point I can live without the things that are still not functioning. 

Thanks for the help.

**Comment by wilpig** [comment]
Display errors causes functions to break because it outputs to the client immediately. You should definitely not have that enabled.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
